glambie/util/timeseries_helpers.py

```diff
@@ -0,0 +1,358 @@
+    """Simple resampling of a an array: returns interpolated y-values based on a new x-array
+
+    Parameters
+    ----------
+    x : np.array
+        The x-values of the input sequence (e.g an array of fractional dates)
+    y : np.array
+        The y-values of the input sequence
+    x_new : np.array
+        The desired x-values of the output sequence (e.g. an array of the new desired time resolution)
+    mode : str, optional
+        the method of interpolation to use, either 'linear', 'nearest' or 'spline'
+        by default "linear"
+
+    Returns
+    -------
+    np.array
+        And array of the y-values at each value in 'x_new'. Will have the same length as x_new
+
+    Raises
+    ------
+    ValueError
+        if mode not recognised
+    """
+    if mode == "spline":
+        s = interpolate.InterpolatedUnivariateSpline(x, y)
+        ynew = s(x_new)
+    elif mode == "linear":
+        ynew = np.interp(x_new, x, y)
+    elif mode == "nearest":
+        s = interpolate.interp1d(x, y, kind='nearest', fill_value="extrapolate")
+        ynew = s(x_new)
+    else:
+        raise ValueError("Unrecognised interpolation mode")
+
+    return ynew
+
+
+def moving_average(dx: float, x: np.array, y: np.array = None, clip: bool = False) -> np.array:
+    """Calculates the moving average of an array.
+
+    If an argument for y is not provided, the function instead provides a moving average of the signal x,
+    over a width of dx samples.
+
+    Parameters
+    ----------
+    dx : float
+        the x-distance over which to average, the window is calculated as 'x > x[i]- (dx/2)' and 'x < x[i] + (dx/2)'
+    x : np.array
+        the x-coordinates of the input data
+    y : np.array, optional
+        the y-coordinates of the input data, by default None
+    clip : bool, optional
+        limit the output data to values for which enough data exists to form a complete average,
+        returned is a np.array of the same size as y with nan values inserted for the clip, by default False
+
+    Returns
+    -------
+    np.array
+        the averaged y-values, has same length as y
+    """
+    if y is not None:
+        n = len(x)
+
+        ry = np.empty(n,)
+        ry.fill(np.NAN)
+
+        for i in range(n):
+            ok = np.logical_and(
+                x > x[i] - dx / 2.,
+                x < x[i] + dx / 2.)
+            if ok.any():
+                ry[i] = np.mean(y[ok])
+        if clip:
+            ok = np.logical_or(
+                x < np.min(x) + dx / 2 - 1,
+                x > np.max(x) - dx / 2 + 1)
+            if ok.any():
+                ry[ok] = np.NAN
+        return ry
+    else:
+        result = np.cumsum(x, dtype=x.dtype)
+        result[dx:] = result[dx:] - result[:-dx]
+        return result[dx - 1:] / dx
+
+
+def timeseries_as_months(fractional_year_array: np.array, downsample_to_month: bool = True) -> np.array(float):
```

Comment:
Very good. I think I'd called `downsample_to_month` something more direct like `fill_in_missing_months` but the current name is fine.

Reply:
sounds good!